### ISSUE TITLE
CSV Importer omits account type when finding account by IBAN

### ISSUE THREAD

**Issue by mattijs79** [firefly-iii/firefly-iii#3426]
**Bug description**
I am running Firefly III version 5.2.6 and CSV Importer version 1.0.9, and my problem is: csv importer doesn't check if it's an expense account or revenue account when finding accounts by IBAN

**Steps to reproduce**
Use csv importer to Import an expense. Then use csv importer to import a deposit FROM THE SAME IBAN. Say you've bought something from a webshop, but you send it back and they return your money. Csv importer finds the expense account, by looking only at the IBAN. When it sends the transaction to firefly it triggers error 'Could not find a valid source account when searching for ID "4" or name ""'

**Comment by JC5** [comment]
Thanks for letting me know! Could you share some log files? The CSV importer should not make that choice when one IBAN results in 2 accounts in Firefly III.

In your log files, something like this should appear:

```
Going to search account with IBAN "X"
Found 2 results, Firefly III will have to make the correct decision.
```

Or, when Firefly III correctly deduces the second transaction is a deposit, this should be the result:

```
Out of cheese error (IBAN)
```

I can do some deep-diving over the weekend.

**Comment by mattijs79** [comment]
The second import:
```
"Datum","Naam / Omschrijving","Rekening","Tegenrekening","Code","Af Bij","Bedrag (EUR)","MutatieSoort","Mededelingen"
"20200326","Horren.com via Mollie","<<<MY_IBAN>>>","<<<WEBSHOP_IBAN>>>","GT","Bij","234,00","Online bankieren","Naam: Horren.com via Mollie Omschrijving: M6280811M1791CU1 0020003173133235 Bestelling 111452 bij Horren.com IBAN:  Kenmerk: 25-03-2020 20:57 0020003173133235 Valutadatum: 25-05-2020"
```

My configuration:
```
{
    "date": "Ymd",
    "default_account": 1,
    "delimiter": "comma",
    "headers": true,
    "ignore_duplicate_lines": true,
    "ignore_duplicate_transactions": true,
    "rules": true,
    "skip_form": true,
    "add_import_tag": true,
    "specifics": [
        "IngDescription"
    ],
    "roles": [
        "date_transaction",
        "opposing-name",
        "account-iban",
        "opposing-iban",
        "_ignore",
        "ing-debit-credit",
        "amount",
        "tags-comma",
        "description",
        "date_book",
        "_ignore"
    ],
    "do_mapping": [
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false
    ],
    "mapping": {
        "1": [],
        "2": [],
        "3": []
    },
    "version": 2
}
```
CSV importer log:
```
[2020-06-04 15:37:33] local.DEBUG: Almost final transaction {"group_title":null,"error_if_duplicate_hash":true,"transactions":[{"type":"withdrawal","date":"2020-03-26 00:00:00","currency_id":null,"currency_code":null,"amount":"234.00","amount_modifier":1,"description":"M6280811M1791CU1 0020003173133235 Bestelling 111452 bij Horren.com Kenmerk: 25-03-2020 20:57 0020003173133235","source_id":null,"source_name":null,"destination_id":null,"destination_name":"Horren.com via Mollie","tags_comma":["Online bankieren"],"tags_space":[],"amount_debit":null,"amount_credit":null,"amount_negated":null,"source_iban":1,"destination_iban":"<<<WEBSHOP_IBAN>>>","book_date":"2020-05-25 00:00:00","original_source":"jc5-csv-import-v1.0.9"}]} 
[2020-06-04 15:37:33] local.INFO: Done parsing and combining 1 lines.  
[2020-06-04 15:37:33] local.DEBUG: Now in App\Services\Import\Routine\PseudoTransactionProcessor::processPseudo  
[2020-06-04 15:37:33] local.INFO: Converting 1 lines into transactions.  
[2020-06-04 15:37:33] local.DEBUG: Now in App\Services\Import\Routine\PseudoTransactionProcessor::processPseudoLine  
[2020-06-04 15:37:33] local.DEBUG: Now running task App\Services\Import\Task\Amount  
[2020-06-04 15:37:33] local.DEBUG: Now at the start of processAmount("234.00")  
[2020-06-04 15:37:33] local.DEBUG: Transaction["amount"] value is not NULL, assume this is the correct value.  
[2020-06-04 15:37:33] local.DEBUG: Amount is now 234.000000000000.  
[2020-06-04 15:37:33] local.DEBUG: Now running task App\Services\Import\Task\Tags  
[2020-06-04 15:37:33] local.DEBUG: Now running task App\Services\Import\Task\Currency  
[2020-06-04 15:37:33] local.DEBUG: Set currency to 1 because it was NULL or empty.  
[2020-06-04 15:37:33] local.DEBUG: Now running task App\Services\Import\Task\Accounts  
[2020-06-04 15:37:33] local.DEBUG: Now in Accounts::process()  
[2020-06-04 15:37:33] local.DEBUG: Now processing transaction 1 of 1  
[2020-06-04 15:37:33] local.DEBUG: Now in Accounts::processTransaction()  
[2020-06-04 15:37:33] local.DEBUG: Now in findAccount {"transaction_type":"withdrawal","id":null,"name":null,"iban":1,"number":null,"bic":null} 
[2020-06-04 15:37:33] local.DEBUG: Default account is #1 ("ING")  
[2020-06-04 15:37:33] local.DEBUG: Find by IBAN.  
[2020-06-04 15:37:33] local.DEBUG: Going to search account with IBAN "1, transactionType: withdrawal"  
[2020-06-04 15:37:33] local.DEBUG: [a] Found expense account #4 based on IBAN "1"  
[2020-06-04 15:37:33] local.DEBUG: Result of findByIBAN is not null, returning: {"id":4,"name":"Horren.com via Mollie","type":"expense","iban":"<<<WEBSHOP_IBAN>>>","number":null,"bic":null,"currency_code":"EUR"} 
[2020-06-04 15:37:33] local.DEBUG: Now in findAccount {"transaction_type":"withdrawal","id":null,"name":"Horren.com via Mollie","iban":"<<<WEBSHOP_IBAN>>>","number":null,"bic":null} 
[2020-06-04 15:37:33] local.DEBUG: findAccount() default account is NULL.  
[2020-06-04 15:37:33] local.DEBUG: Find by IBAN.  
[2020-06-04 15:37:33] local.DEBUG: Going to search account with IBAN "<<<WEBSHOP_IBAN>>>, transactionType: withdrawal"  
[2020-06-04 15:37:33] local.DEBUG: [a] Found expense account #4 based on IBAN "<<<WEBSHOP_IBAN>>>"  
[2020-06-04 15:37:33] local.DEBUG: Result of findByIBAN is not null, returning: {"id":4,"name":"Horren.com via Mollie","type":"expense","iban":"<<<WEBSHOP_IBAN>>>","number":null,"bic":null,"currency_code":"EUR"} 
[2020-06-04 15:37:33] local.DEBUG: Now in determineType("expense", "expense")  
[2020-06-04 15:37:33] local.DEBUG: Check config for "transaction_types.account_to_transaction.expense.expense" and found "". Returning "withdrawal"  
[2020-06-04 15:37:33] local.DEBUG: 234.000000000000 is positive.  
[2020-06-04 15:37:33] local.DEBUG: Now in determineType("expense", "expense")  
[2020-06-04 15:37:33] local.DEBUG: Check config for "transaction_types.account_to_transaction.expense.expense" and found "". Returning "withdrawal"  
[2020-06-04 15:37:33] local.DEBUG: Now running task App\Services\Import\Task\PositiveAmount  
[2020-06-04 15:37:33] local.DEBUG: Now running task App\Services\Import\Task\EmptyDescription  
[2020-06-04 15:37:33] local.DEBUG: Now running task App\Services\Import\Task\EmptyAccounts  
[2020-06-04 15:37:33] local.DEBUG: Now in EmptyAccounts::process()  
[2020-06-04 15:37:33] local.DEBUG: Now processing transaction 1 of 1  
[2020-06-04 15:37:33] local.DEBUG: Now in EmptyAccounts::processTransaction()  
[2020-06-04 15:37:33] local.INFO: Done converting 1 lines into transactions.  
[2020-06-04 15:37:33] local.INFO: Going to submit 1 transactions to your Firefly III instance.  
[2020-06-04 15:37:33] local.INFO: Created tag #11 "CSV Import on 2020-06-04 @ 15:37"  

```
Firefly log
```
[2020-06-04 17:37:33] local.DEBUG: Now in account search()  
[2020-06-04 17:37:33] local.DEBUG: Going to search for "1" in types ["Default account","Cash account","Asset account","Expense account","Revenue account","Initial balance account","Beneficiary account","Import account","Reconciliation account","Loan","Debt","Mortgage"] 
[2020-06-04 17:37:33] local.DEBUG: Found 1 accounts  
[2020-06-04 17:37:33] local.DEBUG: Steam::balance. Native balance is "234.000000000000"  
[2020-06-04 17:37:33] local.DEBUG: Steam::balance. Foreign balance is "0"  
[2020-06-04 17:37:33] local.DEBUG: Steam::balance. Virtual balance is "0"  
[2020-06-04 17:37:33] local.DEBUG: Now in account search()  
[2020-06-04 17:37:33] local.DEBUG: Going to search for "<<<WEBSHOP_IBAN>>>" in types ["Default account","Cash account","Asset account","Expense account","Revenue account","Initial balance account","Beneficiary account","Import account","Reconciliation account","Loan","Debt","Mortgage"] 
[2020-06-04 17:37:33] local.DEBUG: Found 1 accounts  
[2020-06-04 17:37:33] local.DEBUG: Steam::balance. Native balance is "234.000000000000"  
[2020-06-04 17:37:33] local.DEBUG: Steam::balance. Foreign balance is "0"  
[2020-06-04 17:37:33] local.DEBUG: Steam::balance. Virtual balance is "0"  
[2020-06-04 17:37:34] local.DEBUG: Authorize TransactionStoreRequest  
[2020-06-04 17:37:34] local.DEBUG: Collect rules of TransactionStoreRequest  
[2020-06-04 17:37:34] local.DEBUG: Going to validate source_id  
[2020-06-04 17:37:34] local.DEBUG: Now in validateOneTransaction()  
[2020-06-04 17:37:34] local.DEBUG: Added NO errors.  
[2020-06-04 17:37:34] local.DEBUG: Now in GroupValidation::validateDescriptions()  
[2020-06-04 17:37:34] local.DEBUG: Now in validateTransactionTypes()  
[2020-06-04 17:37:34] local.DEBUG: Now in validateForeignCurrencyInformation()  
[2020-06-04 17:37:34] local.DEBUG: Now in validateAccountInformation()  
[2020-06-04 17:37:34] local.DEBUG: Going to loop 1 transaction(s)  
[2020-06-04 17:37:34] local.DEBUG: Transaction type for validator is now Withdrawal  
[2020-06-04 17:37:34] local.DEBUG: Now in AccountValidator::validateSource(4, "", "")  
[2020-06-04 17:37:34] local.DEBUG: Now in validateWithdrawalSource(4, "")  
[2020-06-04 17:37:34] local.DEBUG: Asset account|Loan|Debt|Mortgage  
[2020-06-04 17:37:34] local.WARNING: Not a valid source. Cant find it. ["Asset account","Loan","Debt","Mortgage"] 
[2020-06-04 17:37:34] local.DEBUG: Now in validateEqualAccounts()  
[2020-06-04 17:37:34] local.DEBUG: Now in validateGroupDescription()  
```

**Comment by JC5** [comment]
Thanks. I'm going to release 1.0.10, I hope that solves at least some of the issues people have been getting. Stay tuned!

**Comment by mattijs79** [comment]
Some observations:
* There is a 2 hour time difference between the 2 logs. Perhaps firefly logs in local time and csv importer in UTC?
* This line claims that WEBSHOP IBAN is the destination_iban. It is not, it is the source. 
```
[2020-06-04 15:37:33] local.DEBUG: Almost final transaction {"group_title":null,"error_if_duplicate_hash":true,"transactions":[{"type":"withdrawal","date":"2020-03-26 00:00:00","currency_id":null,"currency_code":null,"amount":"234.00","amount_modifier":1,"description":"M6280811M1791CU1 0020003173133235 Bestelling 111452 bij Horren.com Kenmerk: 25-03-2020 20:57 0020003173133235","source_id":null,"source_name":null,"destination_id":null,"destination_name":"Horren.com via Mollie","tags_comma":["Online bankieren"],"tags_space":[],"amount_debit":null,"amount_credit":null,"amount_negated":null,"source_iban":1,"destination_iban":"<<<WEBSHOP_IBAN>>>","book_date":"2020-05-25 00:00:00","original_source":"jc5-csv-import-v1.0.9"}]} 
```

**Comment by mattijs79** [comment]
I think I found the problem. In CSV importer config/csv_importer.php I find this at line 530: 

```
'opposing-id'           => 'destination_id',
'opposing-iban'         => 'destination_iban',
'opposing-name'         => 'destination_name',
'opposing-number'       => 'destination_number',
'opposing-bic'          => 'destination_bic',
```
This is true for withdrawels, but not for deposits. In that case the opposing-iban is the source. What if you make 2 versions of 'role_to_transaction', one for debit and one for credit, and then use correct one based on the value in 'amount_modifier'?

**Comment by JC5** [comment]
It's easier (only takes one array) to make this conversion regardless of what the source/destination is, and switch them later if necessary.

Firefly III follows "TZ" which is Europe/Amsterdam by default and the CSV importer logs in UTC. I've added a fix :grin:

I've been testing with the latest version (is going to be 1.0.10) which doesn't have this problem. In my test environment, I have an expense account for horren.com with an IBAN. I added a fake IBAN, see below. When searching, this is what happens, which is slightly different from your version:

EE... is my fake IBAN btw.

```
local.DEBUG: findAccount() default account is NULL.
local.DEBUG: Find by IBAN.
local.DEBUG: Going to search account with IBAN "EE483194836736607881"
local.DEBUG: Out of cheese error (IBAN). Found Found expense account #69 based on IBAN "EE483194836736607881". But not going to use expense/deposit combi.
local.DEBUG: Firefly III will have to make the correct decision.
```

My submission is also slightly different (again, fake IBANs):

```
 {
 	"group_title": null,
 	"error_if_duplicate_hash": true,
 	"transactions": [{
 		"type": "deposit",
 		"date": "2020-03-26 00:00:00",
 		"currency_id": 1,
 		"currency_code": null,
 		"amount": "234.000000000000",
 		"amount_modifier": "1",
 		"description": "M6280811M1791CU1 0020003173133235 Bestelling 111452 bij Horren.com IBAN:  Kenmerk: 25-03-2020 20:57 0020003173133235",
 		"source_id": null,
 		"source_name": "Horren.com via Mollie",
 		"destination_id": 1,
 		"destination_name": null,
 		"source_iban": "EE483194836736607881",
 		"destination_iban": null,
 		"book_date": "2020-05-25 00:00:00",
 		"original_source": "jc5-csv-import-v1.0.10",
 		"tags": ["Online bankieren"],
 		"source_number": null,
 		"source_bic": null,
 		"destination_number": null,
 		"destination_bic": null
 	}]
 }
```

So that should be fixed in the next release.

**Comment by mattijs79** [comment]
I'll try again with the next release. Thanks for looking into it.

**Comment by mattijs79** [comment]
Fixed in CSV importer 1.0.10
